src/librustdoc/visit_ast.rs

```diff
@@ -63,23 +63,67 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
     }
 
     pub fn visit(mut self, krate: &'tcx hir::Crate<'_>) -> Module<'tcx> {
-        let mut module = self.visit_mod_contents(
+        let mut top_level_module = self.visit_mod_contents(
             krate.item.span,
             krate.item.attrs,
             &Spanned { span: rustc_span::DUMMY_SP, node: hir::VisibilityKind::Public },
             hir::CRATE_HIR_ID,
             &krate.item.module,
             None,
         );
-        // Attach the crate's exported macros to the top-level module:
-        module
-            .macros
-            .extend(krate.exported_macros.iter().map(|def| self.visit_local_macro(def, None)));
-        module.is_crate = true;
+        top_level_module.is_crate = true;
+        // Attach the crate's exported macros to the top-level module.
+        // In the case of macros 2.0 (`pub macro`), and for built-in `derive`s as well
+        // (_e.g._, `Copy`), these are wrongly bundled in there too, so we need to fix that by
+        // moving them back to their correct locations.
+        krate.exported_macros.iter().for_each(|def| {
+            macro_rules! try_some {($($body:tt)*) => ({
+                fn fn_once<R, F: FnOnce() -> R> (f: F) -> F { f }
+                fn_once(|| Some({ $($body)* }))()
+            })}
```

Comment:
I did hesitate between `try` blocks and this hand-rolled `try_some!` macro a lot, and the thing is, I wanted to have that `is_macro_rules` early escape branch, by early returning `None`. And `try` blocks have no mechanism to early return other than `?`, which means that with a `try` block the code becomes:

```rust
try {
    if def.ast.macro_rules {
        None?;
    }
    ...
}
```

Which can be made to "look" better by being more hacky: `return None?;`, _abusing_ the fact the `return` is actually unreachable 😱

None of these options pleased me, which is the main reason for this being at a draft stage 😞.

If I could understand why the `for` iteration can fail (`pub macro` inside an unnamed item, maybe?), then I could, change the `macro_rules!` branch to a standard one, and then the first `?` could be replaced with a `.map`, propagated through a `.into_iter().flat_map(|it| it.data)`.

___

The other option is to factor that `try` block inside its own function, which would take as parameters what the block / closure currently captures: ` def`, `tcx`, and `&mut top_level_module`. I think that's how people usually circumvent the lack of `try` stuff, since needing such a block to begin with is a sympton of there being enough logic to warrant its own function. In that case, @jyn514:

> where should I put such a helper function?

  - If this were my code /  a crate of mine, I would put it right inside that block, since I like the reduced cognitive burden of super-scoped functions, but I know that most people are not fond of functions scoped inside other function bodies 😅

Reply:
I like a function-scoped function, it's local, does the same thing as `try_some` and is easier to read.